feat(material/core): expose new @use-based Sass API

crisbeto opened the pull request:
Reworks the public Sass API to take advantage of `@use` in order to allow for a single
entry point into `@angular/material` or `@angular/cdk`. Some mixins and variables were
renamed in order be a better reflection of what they do.

Before:
```scss
@import '~@angular/material/theming';
@include mat-core();

$candy-app-primary: mat-palette($mat-indigo);
$candy-app-accent:  mat-palette($mat-pink, A200, A100, A400);
$candy-app-warn:    mat-palette($mat-red);

$candy-app-theme: mat-light-theme((
  color: (
    primary: $candy-app-primary,
    accent: $candy-app-accent,
    warn: $candy-app-warn,
  )
));

@include angular-material-theme($candy-app-theme);
```

After:
```scss
@use '~@angular/material' as mat;
@include mat.core();

$candy-app-primary: mat.define-palette(mat.$indigo-palette);
$candy-app-accent:  mat.define-palette(mat.$pink-palette, A200, A100, A400);
$candy-app-warn:    mat.define-palette(mat.$red-palette);

$candy-app-theme: mat.define-light-theme((
  color: (
    primary: $candy-app-primary,
    accent: $candy-app-accent,
    warn: $candy-app-warn,
  )
));

@include mat.all-component-themes($candy-app-theme);
```

crisbeto commented on src/material/_index.scss:
```diff
@@ -0,0 +1,88 @@
+// Theming APIs
+@forward './core/theming/theming' show define-light-theme, define-dark-theme,
```
Technically we don't _need_ to list out all the APIs with `show`, but I liked being explicit about what we expose. I'm open towards changing it since this could result in us forgetting to re-export a new API.

jelbourn replied:
+1, I like this being explicit as well; from here it's easy to see our entire Sass API surface

crisbeto commented on src/material/core/typography/_typography-utils.scss:
```diff
@@ -12,7 +12,7 @@
 
 // Gets the line height for a level inside a typography config.
-@function height($config, $level) {
+@function line-height($config, $level) {
```
This fixes an accidental rename that happened during the automated migration.

jelbourn reviewed (commented):
Overall amazing, text-field is the only actual comment. I'll start a presubmit now, though, and see what comes up

crisbeto commented on src/dev-app/theme.scss:
```diff
@@ -1,12 +1,14 @@
+@use '../material-experimental/mdc-helpers/focus-indicators-theme' as mdc-focus-indicators-theme;
+@use '../material-experimental/mdc-helpers/focus-indicators' as mdc-focus-indicators;
+@use '../material-experimental/mdc-color/all-color' as mdc-all-color;
+@use '../material-experimental/mdc-theming/all-theme' as mdc-all-theme;
```
It should work, I just didn't want to include more changes than necessary.

crisbeto commented:
@jelbourn I've addressed the feedback by doing the following:
1. Switched a few of our internal themes to use the new API (dev app, e2e app etc). I had to revert my changes to the prebuilt themes, because I couldn't figure out how to format the path correctly. It would have to be something like `@use '../../index' as mat;`, but the compiler complained about it. 
2. I've split up the `text-field` mixin from the CDK into `text-field-autosize` and `text-field-autofill`.

Note that the CI is currently failing, because CircleCI is down. I'll make sure that it's green once Circle is back up.

jelbourn commented on src/material/core/theming/_theming-deprecated.scss:
```diff
@@ -0,0 +1,27 @@
+
+// @deprecated Use `get-color-from-palette`.
+@function color($palette, $hue: default, $opacity: null) {
+  @return get-color-from-palette($palette, $hue, $opacity);
```
This is missing the module name (caught by the presubmit I ran, surprised it didn't fail here)

crisbeto replied:
Fixed.

jelbourn reviewed (approved):
LGTM

jelbourn commented:
Caretaker note: this change might require a build file update. It _shouldn't_ break anything in Google, but let me know if it does. Once it's synced we can follow-up with a change to allow Google apps to use the new APIs (which will require some fun and exciting copybara tweaking)
